control-plane/api-gateway/cache/consul.go

```diff
@@ -339,21 +349,46 @@ func (c *Cache) Write(ctx context.Context, entry api.ConfigEntry) error {
 }
 
 func (c *Cache) ensurePolicy(client *api.Client, gatewayName string) (string, error) {
-	policy := c.gatewayPolicy(gatewayName)
+	c.policyMutex.Lock()
+	defer c.policyMutex.Unlock()
+
+	createPolicy := func() (string, error) {
+		policy := c.gatewayPolicy(gatewayName)
 
-	created, _, err := client.ACL().PolicyCreate(&policy, &api.WriteOptions{})
+		created, _, err := client.ACL().PolicyCreate(&policy, &api.WriteOptions{})
+
+		if isPolicyExistsErr(err, policy.Name) {
+			existing, _, err := client.ACL().PolicyReadByName(policy.Name, &api.QueryOptions{})
+			if err != nil {
+				return "", err
+			}
+			return existing.ID, nil
+		}
 
-	if isPolicyExistsErr(err, policy.Name) {
-		existing, _, err := client.ACL().PolicyReadByName(policy.Name, &api.QueryOptions{})
 		if err != nil {
 			return "", err
 		}
-		return existing.ID, nil
+
+		c.gatewayNameToPolicy[gatewayName] = created
+		return created.ID, nil
 	}
-	if err != nil {
-		return "", err
+
+	cachedPolicy, found := c.gatewayNameToPolicy[gatewayName]
+	if found {
+		existing, _, err := client.ACL().PolicyReadByName(cachedPolicy.Name, &api.QueryOptions{})
+
+		if existing == nil {
+			return createPolicy()
+		}
+
+		if err != nil {
+			return "", err
+		}
+
+		return existing.ID, nil
```

Comment:
Only real suggestion is to make this a bit less nested and more linear like the role code below, it's easier to follow in the below function